panic: assert failed: series key for ID: not found

Running `drop measurement m0` on `tsi`, I hit this panic:

```
panic: runtime error: index out of range

goroutine 360480 [running]:
encoding/binary.binary.bigEndian.Uint16(...)
	/usr/local/go/src/encoding/binary/binary.go:100
github.com/influxdata/influxdb/tsdb.ReadSeriesKeyMeasurement(...)
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/series_file.go:338
github.com/influxdata/influxdb/tsdb.ParseSeriesKey(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc420312000, 0xc420312000, 0x22028b0)
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/series_file.go:359 +0x3b5
github.com/influxdata/influxdb/tsdb.(*seriesIteratorAdapter).Next(0xc448dc8420, 0x0, 0x2710, 0xc447dae000, 0x0)
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/index.go:109 +0xbf
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Engine).DeleteSeriesRange(0xc420368000, 0x1e9eae0, 0xc448dc8420, 0x8000000000000000, 0x7fffffffffffffff, 0x0, 0x0)
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:1190 +0xeb
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Engine).deleteMeasurement(0xc420368000, 0xc436812050, 0x2, 0x8, 0x0, 0x0)
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:1525 +0x1dc
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Engine).DeleteMeasurement(0xc420368000, 0xc436812050, 0x2, 0x8, 0x0, 0xc436812050)
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:1475 +0x67
github.com/influxdata/influxdb/tsdb.(*Shard).DeleteMeasurement(0xc420112000, 0xc436812050, 0x2, 0x8, 0x2, 0x8)
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/shard.go:706 +0x84
github.com/influxdata/influxdb/tsdb.(*Store).DeleteMeasurement.func1(0xc420112000, 0x0, 0x0)
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/store.go:730 +0xd4
github.com/influxdata/influxdb/tsdb.(*Store).walkShards.func1(0xc448dc80e0, 0xc427cf8120, 0xc420112000)
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/store.go:776 +0x48
created by github.com/influxdata/influxdb/tsdb.(*Store).walkShards
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/store.go:775 +0x87
```

After restarting the server, I then hit this:

```
panic: assert failed: series key for ID: 2606674 not found

goroutine 84 [running]:
github.com/influxdata/influxdb/tsdb/index/tsi1.assert(0x19ede00, 0xc42017e0f0, 0x24, 0x0, 0x0, 0x0)
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/tsi1.go:531 +0xf6
github.com/influxdata/influxdb/tsdb/index/tsi1.(*LogFile).execSeriesEntry(0xc420396240, 0xc420b9bc60)
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/log_file.go:622 +0x225
github.com/influxdata/influxdb/tsdb/index/tsi1.(*LogFile).execEntry(0xc420396240, 0xc420b9bc60)
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/log_file.go:576 +0x68
github.com/influxdata/influxdb/tsdb/index/tsi1.(*LogFile).open(0xc420396240, 0xc42018e640, 0xc42018e5f0)
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/log_file.go:135 +0x3a8
github.com/influxdata/influxdb/tsdb/index/tsi1.(*LogFile).Open(0xc420396240, 0xc42357e000, 0x44)
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/log_file.go:83 +0x2b
github.com/influxdata/influxdb/tsdb/index/tsi1.(*Partition).openLogFile(0xc42040ca20, 0xc42357e000, 0x44, 0xc42357e000, 0x44, 0x8)
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/partition.go:222 +0x5a
github.com/influxdata/influxdb/tsdb/index/tsi1.(*Partition).Open(0xc42040ca20, 0x0, 0x0)
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/partition.go:164 +0x5f4
github.com/influxdata/influxdb/tsdb/index/tsi1.(*Index).Open.func1(0xc421046104, 0x8, 0xc420194000, 0xc4204b6780, 0x2)
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/index.go:195 +0x30
created by github.com/influxdata/influxdb/tsdb/index/tsi1.(*Index).Open
	/Users/jason/go/src/github.com/influxdata/influxdb/tsdb/index/tsi1/index.go:189 +0x640
```

--- a/tsdb/index.go
+++ b/tsdb/index.go
@@ -99,22 +99,30 @@ type seriesIteratorAdapter struct {
 func (itr *seriesIteratorAdapter) Close() error { return itr.itr.Close() }
 
 func (itr *seriesIteratorAdapter) Next() (SeriesElem, error) {
-	elem, err := itr.itr.Next()
-	if err != nil {
-		return nil, err
-	} else if elem.SeriesID == 0 {
-		return nil, nil
+	for {
+		elem, err := itr.itr.Next()
+		if err != nil {
+			return nil, err
+		} else if elem.SeriesID == 0 {
+			return nil, nil
+		}
+
+		// Skip if this key has been tombstoned.
+		key := itr.sfile.SeriesKey(elem.SeriesID)
+		if len(key) == 0 {
+			continue
+		}
+
+		name, tags := ParseSeriesKey(key)
+		deleted := itr.sfile.IsDeleted(elem.SeriesID)
+
+		return &seriesElemAdapter{
+			name:    name,
+			tags:    tags,
+			deleted: deleted,
+			expr:    elem.Expr,
+		}, nil
 	}
-
-	name, tags := ParseSeriesKey(itr.sfile.SeriesKey(elem.SeriesID))
-	deleted := itr.sfile.IsDeleted(elem.SeriesID)
-
-	return &seriesElemAdapter{
-		name:    name,
-		tags:    tags,
-		deleted: deleted,
-		expr:    elem.Expr,
-	}, nil
 }
 
 type seriesElemAdapter struct {
@@ -237,26 +245,34 @@ func (itr *seriesQueryAdapterIterator) Close() error {
 
 // Next emits the next point in the iterator.
 func (itr *seriesQueryAdapterIterator) Next() (*query.FloatPoint, error) {
-	// Read next series element.
-	e, err := itr.itr.Next()
-	if err != nil {
-		return nil, err
-	} else if e.SeriesID == 0 {
-		return nil, nil
-	}
-
-	// Convert to a key.
-	name, tags := ParseSeriesKey(itr.sfile.SeriesKey(e.SeriesID))
-	key := string(models.MakeKey(name, tags))
-
-	// Write auxiliary fields.
-	for i, f := range itr.opt.Aux {
-		switch f.Val {
-		case "key":
-			itr.point.Aux[i] = key
+	for {
+		// Read next series element.
+		e, err := itr.itr.Next()
+		if err != nil {
+			return nil, err
+		} else if e.SeriesID == 0 {
+			return nil, nil
 		}
+
+		// Skip if key has been tombstoned.
+		seriesKey := itr.sfile.SeriesKey(e.SeriesID)
+		if len(seriesKey) == 0 {
+			continue
+		}
+
+		// Convert to a key.
+		name, tags := ParseSeriesKey(seriesKey)
+		key := string(models.MakeKey(name, tags))
+
+		// Write auxiliary fields.
+		for i, f := range itr.opt.Aux {
+			switch f.Val {
+			case "key":
+				itr.point.Aux[i] = key
+			}
+		}
+		return &itr.point, nil
 	}
-	return &itr.point, nil
 }
 
 // filterUndeletedSeriesIDIterator returns all series which are not deleted.
@@ -744,7 +760,12 @@ func (itr *seriesPointIterator) readSeriesKeys(name []byte) error {
 		} else if elem.SeriesID == 0 {
 			break
 		}
-		itr.keys = append(itr.keys, itr.indexSet.SeriesFile.SeriesKey(elem.SeriesID))
+
+		key := itr.indexSet.SeriesFile.SeriesKey(elem.SeriesID)
+		if len(key) == 0 {
+			continue
+		}
+		itr.keys = append(itr.keys, key)
 	}
 
 	// Sort keys.
@@ -1719,7 +1740,9 @@ func (is IndexSet) MeasurementSeriesKeysByExpr(name []byte, expr influxql.Expr) 
 		}
 
 		seriesKey := is.SeriesFile.SeriesKey(e.SeriesID)
-		assert(seriesKey != nil, fmt.Sprintf("series key for ID: %d not found", e.SeriesID))
+		if len(seriesKey) == 0 {
+			continue
+		}
 
 		name, tags := ParseSeriesKey(seriesKey)
 		keys = append(keys, models.MakeKey(name, tags))
@@ -2145,7 +2168,7 @@ func (is IndexSet) tagValuesByKeyAndExpr(auth query.Authorizer, name []byte, key
 		}
 
 		buf := is.SeriesFile.SeriesKey(e.SeriesID)
-		if buf == nil {
+		if len(buf) == 0 {
 			continue
 		}
 
@@ -2305,7 +2328,13 @@ func (is IndexSet) TagSets(sfile *SeriesFile, name []byte, opt query.IteratorOpt
 				break
 			}
 
-			_, tags := ParseSeriesKey(sfile.SeriesKey(e.SeriesID))
+			// Skip if the series has been tombstoned.
+			key := sfile.SeriesKey(e.SeriesID)
+			if len(key) == 0 {
+				continue
+			}
+
+			_, tags := ParseSeriesKey(key)
 			if opt.Authorizer != nil && !opt.Authorizer.AuthorizeSeriesRead(is.Database(), name, tags) {
 				continue
 			}

